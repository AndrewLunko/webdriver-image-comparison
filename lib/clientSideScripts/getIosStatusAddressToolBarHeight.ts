import {StatusAddressToolBarHeight} from './statusAddressToolBarHeight.interfaces';
import {IosOffsets} from '../helpers/constants.interfaces';

/**
 * Get the current height of the iOS status and address bar
 */
export default function getIosStatusAddressToolBarHeight(iosOffsets: IosOffsets): StatusAddressToolBarHeight {
  // Determine version for the right offsets
  const match = (navigator.appVersion).match(/OS (\d+)_(\d+)_?(\d+)?/);
  const majorVersion = parseInt(match[1], 10);
  const versionOffsets = iosOffsets[majorVersion];

  // Determine screen widht/height to determine iPhone X series address bar height
  const {width, height} = window.screen;
  const isIphoneXSeries = (width === 812 || height === 812) || (width === 896 || height === 896);

  // Determine if iPad Pro with no touchID for higher status bar
  const {innerWidth, innerHeight} = window;
  const isIpadNoTouch = (
    (innerHeight + versionOffsets.ADDRESS_BAR + versionOffsets.STATUS_BAR_PRO) === height ||
    (innerWidth + versionOffsets.ADDRESS_BAR + versionOffsets.STATUS_BAR_PRO) === width
  );

  // Determine address bar height
  let statusBarHeight;
  if (isIphoneXSeries) {
    statusBarHeight = versionOffsets.STATUS_BAR_X;
  } else if (isIpadNoTouch) {
    statusBarHeight = versionOffsets.STATUS_BAR_PRO;
  } else {
    statusBarHeight = versionOffsets.STATUS_BAR;
  }

  const statusAddressBarHeight = statusBarHeight + versionOffsets.ADDRESS_BAR;
  const toolBarHeight = height - innerHeight - statusAddressBarHeight;
  const toolBar = {
    height: toolBarHeight,
    width,
    x: 0,
    y: height - toolBarHeight,
  };

  // The status and address can be smaller due to a "manual" scroll meaning there is no toolbar
  if (toolBar.height < 0) {
    toolBar.height = 0;
    toolBar.y = 0;
  }
  // Add the handlebar on iPhone X series as a blockout, it can colour with the background
  if (toolBar.height === 0 && isIphoneXSeries) {
    // Needs to be verbose to support the correct injection
    toolBar.height = versionOffsets.HOME_BAR.DEFAULT.height;
    toolBar.width = versionOffsets.HOME_BAR.DEFAULT.width;
    toolBar.x = versionOffsets.HOME_BAR.DEFAULT.x;
    toolBar.y = versionOffsets.HOME_BAR.DEFAULT.y;

    // For the XS Max
    if ((width === 896 || height === 896)) {
      // Needs to be verbose to support the correct injection
      toolBar.height = versionOffsets.HOME_BAR.LARGE.height;
      toolBar.width = versionOffsets.HOME_BAR.LARGE.width;
      toolBar.x = versionOffsets.HOME_BAR.LARGE.x;
      toolBar.y = versionOffsets.HOME_BAR.LARGE.y;
    }
  }

  // Determine status, address and tool bar height and width
  return {
    statusAddressBar: {
      height: statusAddressBarHeight,
      width,
      x: 0,
      y: 0,
    },
    toolBar,
  };
}
