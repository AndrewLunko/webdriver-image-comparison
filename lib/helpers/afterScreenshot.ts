import hideScrollBars from '../clientSideScripts/hideScrollbars';
import removeCustomCss from '../clientSideScripts/removeCustomCss';
import {CUSTOM_CSS_ID} from './constants';
import {formatFileName, getAndCreatePath} from './utils';
import {saveBase64Image} from '../methods/images';
import {join} from 'path';
import {Executor} from '../methods/methods.interface';
import {AfterScreenshotOptions, ScreenshotOutput} from './afterScreenshot.interfaces';
import hideRemoveElements from '../clientSideScripts/hideRemoveElements';

/**
 * Methods that need to be executed after a screenshot has been taken
 * to set all back to the original state
 */
export default async function afterScreenshot(executor: Executor, options: AfterScreenshotOptions): Promise<ScreenshotOutput> {
  const {
    actualFolder,
    base64Image,
    fileName: fileNameOptions,
    filePath,
    hideElements,
    hideScrollBars: noScrollBars,
    removeElements,
  } = options;

  // Get the path
  const path = getAndCreatePath(actualFolder, filePath);

  // Get the filePath
  const fileName = formatFileName(fileNameOptions);

  // Save the element
  await saveBase64Image(base64Image, join(path, fileName));

  // Show the scrollbars again
  await executor(hideScrollBars, !noScrollBars);

  // Show elements again
  if (hideElements.length > 0 || removeElements.length > 0) {
    await executor(hideRemoveElements, {hide: hideElements, remove: removeElements}, false);
  }

  // Remove the custom set css
  await executor(removeCustomCss, CUSTOM_CSS_ID);

  // Return the needed data
  return {
    fileName,
    path,
    devicePixelRatio: fileNameOptions.devicePixelRatio,
  };
}
