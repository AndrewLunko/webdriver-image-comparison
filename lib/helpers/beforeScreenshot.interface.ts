import {EnrichedInstanceData, InstanceData} from '../methods/instanceData.interfaces';

export interface BeforeScreenshotOptions {
  // The instance data
  instanceData: InstanceData;
  // The padding that needs to be added to the address bar on iOS and Android to do a proper cutout of the the viewport.
  addressBarShadowPadding: number;
  // Disable all css animations
  disableCSSAnimation: boolean;
  // Hide all scrollbars
  noScrollBars: boolean;
  // The padding that needs to be added to the tool bar on iOS and Android
  toolBarShadowPadding: number;
}

export interface BeforeScreenshotResult extends EnrichedInstanceData {};
