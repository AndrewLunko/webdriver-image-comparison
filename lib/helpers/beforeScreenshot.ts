import hideScrollBars from '../clientSideScripts/hideScrollbars';
import setCustomCss from '../clientSideScripts/setCustomCss';
import {CUSTOM_CSS_ID} from './constants';
import {getAddressBarShadowPadding, getToolBarShadowPadding} from './utils';
import getEnrichedInstanceData from '../methods/instanceData';
import {BeforeScreenshotOptions, BeforeScreenshotResult} from './beforeScreenshot.interface';
import {Executor} from '../methods/methods.interface';
import hideRemoveElements from '../clientSideScripts/hideRemoveElements';

/**
 * Methods that need to be executed before a screenshot will be taken
 */
export default async function beforeScreenshot(
  executor: Executor,
  options: BeforeScreenshotOptions,
  addShadowPadding: boolean = false
): Promise<BeforeScreenshotResult> {

  const {browserName, nativeWebScreenshot, platformName} = options.instanceData;
  const {
    addressBarShadowPadding,
    disableCSSAnimation,
    hideElements,
    noScrollBars,
    removeElements,
    toolBarShadowPadding,
  } = options;
  const addressBarPadding = getAddressBarShadowPadding({
    platformName,
    browserName,
    nativeWebScreenshot,
    addressBarShadowPadding,
    addShadowPadding,
  });
  const toolBarPadding = getToolBarShadowPadding({platformName, browserName, toolBarShadowPadding, addShadowPadding});

  // Hide the scrollbars
  await executor(hideScrollBars, noScrollBars);

  // Hide and or Remove elements
  if (hideElements.length > 0 || removeElements.length > 0) {
    await executor(hideRemoveElements, {hide: hideElements, remove: removeElements}, true);
  }

  // Set some custom css
  await executor(setCustomCss, {addressBarPadding, disableCSSAnimation, id: CUSTOM_CSS_ID, toolBarPadding});

  // Get all the needed instance data
  const instanceOptions = {
    addressBarShadowPadding: options.addressBarShadowPadding,
    toolBarShadowPadding: options.toolBarShadowPadding,
    ...(options.instanceData),
  };

  return getEnrichedInstanceData(executor, instanceOptions, addShadowPadding);
}
