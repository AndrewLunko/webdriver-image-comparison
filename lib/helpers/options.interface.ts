export interface ClassOptions{
  /**
   * Class options
   */
  addressBarShadowPadding?: number;
  autoSaveBaseline?: boolean;
  baselineFolder?: string;
  debug?: boolean;
  formatImageName?: string;
  nativeWebScreenshot?: boolean;
  savePerInstance?: boolean;
  screenshotPath?: string;
  toolBarShadowPadding?: number;

  /**
   * Class and method options
   */
  disableCSSAnimation?: boolean;
  fullPageScrollTimeout?: number;
  hideScrollBars?: boolean;

  /**
   * Compare options
   */
  blockOutStatusBar?: boolean;
  blockOutToolBar?: boolean;
  ignoreAlpha?: boolean;
  ignoreAntialiasing?: boolean;
  ignoreColors?: boolean;
  ignoreLess?: boolean;
  ignoreNothing?: boolean;
  ignoreTransparentPixel?: boolean;
  rawMisMatchPercentage?: boolean;
  returnAllCompareData?: boolean;
  saveAboveTolerance?: number;
}

export interface DefaultOptions {
  addressBarShadowPadding: number;
  autoSaveBaseline: boolean;
  debug: boolean;
  formatImageName: string;
  nativeWebScreenshot: boolean;
  savePerInstance: boolean;
  toolBarShadowPadding: number;
  resizeDimensions: {
    bottom: number;
    left: number;
    right: number;
    top: number;
  };
  disableCSSAnimation: boolean;
  fullPageScrollTimeout: number;
  hideScrollBars: boolean;
  compareOptions: CompareOptions
}

interface CompareOptions {
  blockOutStatusBar: boolean;
  blockOutToolBar: boolean;
  ignoreAlpha: boolean;
  ignoreAntialiasing: boolean;
  ignoreColors: boolean;
  ignoreLess: boolean;
  ignoreNothing: boolean;
  ignoreTransparentPixel: boolean;
  rawMisMatchPercentage: boolean;
  returnAllCompareData: boolean;
  saveAboveTolerance: number;
}
