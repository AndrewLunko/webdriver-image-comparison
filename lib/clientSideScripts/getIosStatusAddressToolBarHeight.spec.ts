import {IOS_DEVICES, NAVIGATOR_APP_VERSIONS} from '../../mocks/mocks';
import getIosStatusAddressToolBarHeight from './getIosStatusAddressToolBarHeight';
import {OFFSETS} from '../helpers/constants';

const CONFIGURABLE = {
  writable: true,
  configurable: true,
};

describe('getIosStatusAddressToolBarHeight', () => {
  it('should get the correct status, address and toolbar height for an iPhone with iOS 10', () => {
    setEnvironment(10, 'IPHONE');

    expect(getIosStatusAddressToolBarHeight(OFFSETS.IOS)).toMatchSnapshot();
  });

  it('should get the correct toolbar height for an iPhone with iOS 12 that has been scrolled', () => {
    setEnvironment(12, 'IPHONE');
    Object.defineProperty(window, 'innerHeight', {value: 598, ...CONFIGURABLE});

    expect(getIosStatusAddressToolBarHeight(OFFSETS.IOS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for an iPhone X with iOS 11', () => {
    setEnvironment(11, 'IPHONE_X');

    expect(getIosStatusAddressToolBarHeight(OFFSETS.IOS)).toMatchSnapshot();
  });

  it('should get the correct homebar as toolbar height for an iPhone X with iOS 12 that has been scrolled', () => {
    setEnvironment(12, 'IPHONE_X');
    Object.defineProperty(window, 'innerHeight', {value: 718, ...CONFIGURABLE});

    expect(getIosStatusAddressToolBarHeight(OFFSETS.IOS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for an iPhone XS Max with iOS 12', () => {
    setEnvironment(12, 'IPHONE_XS_MAX');

    expect(getIosStatusAddressToolBarHeight(OFFSETS.IOS)).toMatchSnapshot();
  });

  it('should get the correct homebar as toolbar height for an iPhone XS MAx with iOS 12 that has been scrolled', () => {
    setEnvironment(12, 'IPHONE_XS_MAX');
    Object.defineProperty(window, 'innerHeight', {value: 802, ...CONFIGURABLE});

    expect(getIosStatusAddressToolBarHeight(OFFSETS.IOS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for an iPad with iOS 12', () => {
    setEnvironment(12, 'IPAD');

    expect(getIosStatusAddressToolBarHeight(OFFSETS.IOS)).toMatchSnapshot();
  });

  it('should get the correct status, address and toolbar height for an iPad Pro with to touch id and with iOS 12', () => {
    setEnvironment(12, 'IPAD_NO_TOUCH');

    expect(getIosStatusAddressToolBarHeight(OFFSETS.IOS)).toMatchSnapshot();
  });
});


/**
 * Set the environment for the test
 */
function setEnvironment(ios:number, phone: string){
  // @ts-ignore
  Object.defineProperty(navigator, 'appVersion', {value: NAVIGATOR_APP_VERSIONS.IOS[ios], ...CONFIGURABLE});
  // @ts-ignore
  Object.defineProperty(window.screen, 'width', {value: IOS_DEVICES[phone].width, ...CONFIGURABLE});
  // @ts-ignore
  Object.defineProperty(window.screen, 'height', {value: IOS_DEVICES[phone].height, ...CONFIGURABLE});
  // @ts-ignore
  Object.defineProperty(window, 'innerWidth', {value: IOS_DEVICES[phone].innerWidth, ...CONFIGURABLE});
  // @ts-ignore
  Object.defineProperty(window, 'innerHeight', {value: IOS_DEVICES[phone].innerHeight, ...CONFIGURABLE});
}
