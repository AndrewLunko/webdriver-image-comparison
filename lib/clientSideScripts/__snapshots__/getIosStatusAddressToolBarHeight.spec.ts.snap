// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIosStatusAddressToolBarHeight should get the correct homebar as toolbar height for an iPhone X with iOS 12 that has been scrolled 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 94,
    "width": 375,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 5,
    "width": 135,
    "x": 120,
    "y": 799,
  },
}
`;

exports[`getIosStatusAddressToolBarHeight should get the correct homebar as toolbar height for an iPhone XS MAx with iOS 12 that has been scrolled 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 94,
    "width": 414,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 5,
    "width": 148,
    "x": 133,
    "y": 883,
  },
}
`;

exports[`getIosStatusAddressToolBarHeight should get the correct status, address and toolbar height for an iPad Pro with to touch id and with iOS 12 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 74,
    "width": 1024,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 0,
    "width": 1024,
    "x": 0,
    "y": 1366,
  },
}
`;

exports[`getIosStatusAddressToolBarHeight should get the correct status, address and toolbar height for an iPad with iOS 12 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 70,
    "width": 768,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 0,
    "width": 768,
    "x": 0,
    "y": 1024,
  },
}
`;

exports[`getIosStatusAddressToolBarHeight should get the correct status, address and toolbar height for an iPad with iPadOS 13 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 70,
    "width": 768,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 0,
    "width": 768,
    "x": 0,
    "y": 1024,
  },
}
`;

exports[`getIosStatusAddressToolBarHeight should get the correct status, address and toolbar height for an iPhone X with iOS 11 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 94,
    "width": 375,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 83,
    "width": 375,
    "x": 0,
    "y": 729,
  },
}
`;

exports[`getIosStatusAddressToolBarHeight should get the correct status, address and toolbar height for an iPhone XS Max with iOS 12 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 94,
    "width": 414,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 83,
    "width": 414,
    "x": 0,
    "y": 813,
  },
}
`;

exports[`getIosStatusAddressToolBarHeight should get the correct status, address and toolbar height for an iPhone with iOS 10 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 64,
    "width": 375,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 50,
    "width": 375,
    "x": 0,
    "y": 617,
  },
}
`;

exports[`getIosStatusAddressToolBarHeight should get the correct toolbar height for an iPhone with iOS 12 that has been scrolled 1`] = `
Object {
  "statusAddressBar": Object {
    "height": 70,
    "width": 375,
    "x": 0,
    "y": 0,
  },
  "toolBar": Object {
    "height": 0,
    "width": 375,
    "x": 0,
    "y": 0,
  },
}
`;
